rs/execution_environment/src/history.rs

```diff
@@ -159,8 +181,60 @@ impl IngressHistoryWriter for IngressHistoryWriterImpl {
                 status
             );
         }
+
         use IngressState::*;
         use IngressStatus::*;
+
+        // Computes the duration elapsed since the time when a message was included into
+        // a block (approximated by block time); and since its induction.
+        let message_latencies_seconds = |message_id| {
+            let received_time = self.received_time.read().unwrap();
```

Comment:
Is `set_status` ever called in a context where we hold other locks so that this could end up leading to a deadlock?

Reply:
You only get a deadlock if you acquire two locks in different order. This is the only lock in `IngressHistoryWriterImpl` and we don't call out to anywhere (where there may be a second lock that might be held by another thread).

Reply:
Yeah, deadlock was not the right word. What I had in mind was a scenario where we already hold the the lock and then call this function, so it would be a rather trivial deadlock where a thread locks itself. But I assume there is also not anything like this?

Reply:
Yeah, we don't make any calls (except bumping metrics) while holding the lock, AFAICT.

And the lock isn't even intended to protect anything, as only the Batch Processor thread will ever access it. It's only there to make Rust happy regarding interior mutability (i.e. so that methods that take `&self` can mutate `received_time`). Which seems to be necessary because we pass `Arc<dyn IngressHistoryWriter>` clones everywhere throughout Execution (but there's obviously a single state they can write to, so there cannot be any concurrent calls).